Sg 13282 python3

pscadding opened the pull request:
Added Python 3 support as well as PySide 2 support for the console app when used outside of Toolkit.

Also added some tests using the pytest framework. The coverage is likely to be fairly small, but it covers some of the basic functionality.

Updated to require `tk-framework-shotgunutils` `v5.x.x` due to Python 3 support only being in v5.

pscadding committed:
ran python-modernize
Now compiles with python3.
Also added black and hound badges to the readme.

pscadding committed:
Added the pre commit hook config, and ran the pre commit hook.
Contains the black reformatting.
Marked the README.md, LICENSE, .gitignore and python/app/console.py
files as not executables.

pscadding committed:
Added azure pipelines config file.

pscadding committed:
added skip tests parameter to the azure config.

pscadding committed:
Added Azure status badge to the readme

pscadding committed:
Addressing some of the Hound CI comments.

pscadding committed:
Fixing Azure badge formatting in the README

pscadding committed:
Adjusting the order of the README badges

pscadding committed:
fixing badge order again.

pscadding committed:
Python 3 updates

Fixed a bug with the build resources script not creating new lines
in the modified QT import section.

Bumped `tk-framework-shotgunutils` to version 5 to support python3

There are a couple of TODOs outstanding.

pscadding committed:
Reversed temporary commented out line

pscadding committed:
removed the skipping of tests now it should be python 3 compatible.

pscadding committed:
Bug fixes, PySide2 fix, and tidy up.

- Fixes a bug with showing the open file dialog, when the path is None.
- Removed unused imports in the external_sources.py hook.
- Python3 compatibility for the external_sources_gists.py hook example.

pscadding committed:
Made checked kwarg an arg instead.

pscadding committed:
Collection of small fixes.
Fixed the unicode issues around printing the output
Implemented try except around importing six in non TK specific code.
Fixed Lambda definition for PySide.

pscadding committed:
addressing Hound CI comments

pscadding committed:
Merge branch 'black' into SG-13282-python3

pscadding committed:
Added a qt importer to handle importing from sgtk, PySide and PySide2.

pscadding committed:
removed dependency on sgtk

pscadding committed:
Added very basic tests.
Checks that sgtk can't be imported for a standalone test.
Checks that the console app can be imported.

pscadding committed:
Added more tests.

pscadding committed:
Added the python badge to the repo.
I've not included 2.6 as I can't test it on my machine at the moment,
and the automated tests don't cover it.

pscadding committed:
Added additional test scripts, and now checks execute output.

pscadding committed:
added a comment

pscadding committed:
addressed Hound comment.

pscadding committed:
Merge branch 'master' into SG-13282-python3

# Conflicts:
#	README.md
#	azure-pipelines.yml
#	hooks/external_sources_gists.py
#	resources/build_resources.sh

pscadding commented on python/app/console.py:
```diff
@@ -222,7 +221,7 @@ def open(self, path=None):
                 parent=QtGui.QApplication.activeWindow(),
                 caption="Open Python Script",
-                directory=path,
+                directory=path if path else "",
```
Small bugfix here, where it would fail if the path was None.

pscadding commented on python/app/input_widget.py:
```diff
@@ -184,7 +181,7 @@ def execute(self):
 
         # exec
         else:
-            with nested(self._stdout_redirect, self._stdin_redirect):
```
I can't test on Python 2, but I'm concerned this might break it?

jfboismenu replied:
Why did you remove the nested call? I thought this was introduced in Python 2.6 because you can't have multiple values with a single `with` statement IIRC?

pscadding replied:
https://stackoverflow.com/questions/893333/multiple-variables-in-a-with-statement
Python 3 doesn't support `nested`, instead the approach is to just provide a tuple.

jfboismenu replied:
Could you nest the calls with two `with` then, so as to preserve 2.6 compatibility? I'm sure Francis will find a way to test it.

pscadding replied:
good idea!

pscadding commented on python/app/input_widget.py:
```diff
@@ -439,8 +436,7 @@ def _current_line_color(self):
     def _format_exc(self):
         """Get the latest stack trace and format it for display."""
-        tb = sys.exc_info()[2]
-        return traceback.format_exc(tb)
```
Bugfix: I don't think the traceback should ever have been passed through to this method as the `limit` kwarg, although it doesn't error in Python 2 as it seems the `>=` operator can be used with the traceback object in Python 2.

jfboismenu replied:
I'm confused. What are you saying here exactly?

pscadding replied:
[`traceback.format_exec`](https://docs.python.org/2/library/traceback.html#traceback.format_exc) only supports a limit kwarg. We were passing a traceback object through as if it was a limit, and that seemed to work in Python 2. The traceback object could be used with a less or more than operator. In Python 3 it can't, but I think this was a bug anway in the original code anyway.

jfboismenu replied:
oh, get it!

pscadding committed:
Updated some comments.

pscadding committed:
Fixed a Windows unicode error.

pscadding committed:
Switched to using io.open so it can handle the encoding in 2.7 as well.

pscadding committed:
Adjusted test to handle opening the file the same way the app does.

jfboismenu commented on README.md:
```diff
@@ -1,3 +1,4 @@
+[![Python 2.7 3.7](https://img.shields.io/badge/python-2.6%20%7C%202.7%20%7C%203.7-blue.svg)](https://www.python.org/)
```
Missing 2.6 in the second pair of `[]`. This is for the tooltip IIRC.

pscadding replied:
So I removed the 2.6 because I have no way of testing it at the moment, and our automated tests don't check py 2.6.
Maybe one for in discussion, I spent a short while trying to figure out how to install py 2.6, but I'm not sure if it's worth the time?

jfboismenu commented on azure-pipelines.yml:
```diff
@@ -30,5 +30,3 @@ variables:
 # Launch into the build pipeline.
 jobs:
 - template: build-pipeline.yml@templates
-  parameters:
```
I love this :)

jfboismenu commented on python/app/console.py:
```diff
@@ -236,19 +238,20 @@ def open(self, path=None):
         # clear the contents, open and load the file
-        fh = open(path)
-        try:
+        with io.open(path, "r", encoding="utf-8") as fh:
```
It will be interesting to see if this becomes a pattern for us in the future, having to use `io.open`.

jfboismenu replied:
Could you leave a comment in the code explaining why you need to do this instead of simply calling `open`. Maybe also leave a FIXME comment saying it makes things better, but not quite perfect for unicode strings.

pscadding replied:
Sounds good!

jfboismenu commented on python/app/redirect.py:
```diff
@@ -101,9 +99,6 @@ def write(self, msg):
 
         If tee, then also write to stdout.
         """
-        if isinstance(msg, unicode):
```
This is a case where it would make sense to put a comment in the PR explaning why we are removing code.

pscadding replied:
Good point!
So we are removing this line as it is not compatible with python 3, and this module is designed to run without sgtk being present so we can't rely on the existence of `six` to smooth things over.

I was decided however it was better to add a check to the check to the `OutputStreamWidget.add_output` as this was where it was going to get added to the UI, so we should sanitize it there. It still can't rely on six being there, but it will try to use it if six could be imported.

jfboismenu commented on tests/test_standalone_app.py:
```diff
@@ -0,0 +1,163 @@
+# -*- coding: utf-8 -*-
```
Missing copyright.

jfboismenu commented on tests/test_standalone_app.py:
```diff
@@ -0,0 +1,163 @@
+
+    # make sure tk-core is removed from the path so that
+    # we can test the python console app standalone.
+    for a_path in reversed(sys.path):
```
It's removed from the path, but how are you sure it is not already imported?

pscadding replied:
If it's removed from the path, and I have a test making sure it can't be imported is that not enough?
If the app can't import sgtk then it shouldn't be able to use it.

jfboismenu replied:
Oh, right, I missed that test, sorry!

pscadding replied:
No worries!

jfboismenu commented on tests/test_standalone_app.py:
```diff
@@ -0,0 +1,163 @@
+    make sure we can't import sgtk from tk-core.
+    :return:
+    """
+    print(sys.path)
```
Stray print?

jfboismenu commented on tests/test_standalone_app.py:
```diff
@@ -0,0 +1,163 @@
+    assert console_widget.tabs.count() == 1
+
+
+@pytest.mark.parametrize(
```
❤️ !!!

jfboismenu commented on tests/test_standalone_app.py:
```diff
@@ -0,0 +1,163 @@
+    "script, python_version",
+    [
+        ("resource_script.py", 2),
+        # only compare the contents of this one in Python 3
```
Why are we running some tests only in Python 3 and others in both? Seems like a comment is in order.

pscadding replied:
Good point.
I've actually removed the check in the open script test, as it is no longer needed.

The execute test still yields different results depending on the Python version, so since I was checking that the output is what I would expect, I opted to only test the unicode tests in Python 3 where the output is what I would expect. Python 2 doesn't error during execution but the output is not the same as the input due to the way it translates the unicode.

I've added a comment as you suggested.

jfboismenu reviewed (changes requested):
Looks good, but I do have a couple of questions.

pscadding committed:
Added codecov badge

jfboismenu commented on python/app/qt_importer.py:
```diff
@@ -0,0 +1,37 @@
+# NOTE: This repo is typically used as a Toolkit app, but it is also possible use the console in a
```
Missing copyright!

pscadding replied:
Good catch, I'm not used to writing new modules yet :)

jfboismenu commented on tests/test_standalone_app.py:
```diff
@@ -0,0 +1,163 @@
+    """
+    print(sys.path)
+    with pytest.raises(ImportError):
+        import sgtk
```
# noqa

jfboismenu replied:
or maybe `import sgtk as _` ?

pscadding committed:
Addressing CR comments, and some small tweaks.
Added a fall back to convert to str in the output widget if six is not
present.
I don't think the python 3 vs 2 separation is needed in the open script
test anymore so I have removed it.

pscadding committed:
Added some comments around the Python 2v3 testing separation.
Also slightly reordered the testing list, so that the Python 3 tests
were grouped together at the bottom.

pscadding committed:
Updated the python badge in README to include 2.6.
I will get someone to test 2.6.

Also addressed a hound comment.

jfboismenu commented on python/app/input_widget.py:
```diff
@@ -171,26 +168,32 @@ def execute(self):
-                        stderr.write(self._format_exc())
-                else:
-                    self.results.emit(str(results))
+            # Use two with statements inside each other as python 2.6 doesn't support, passing a tuple
```
```suggestion
            # Use two with statements inside each other as python 2.6 doesn't support passing a tuple
```

jfboismenu commented on python/app/input_widget.py:
```diff
@@ -171,26 +168,32 @@ def execute(self):
-                    # oops, error encountered. write/redirect to the error signal
-                    with self._stderr_redirect as stderr:
-                        stderr.write(self._format_exc())
+            # Use two with statements inside each other as python 2.6 doesn't support, passing a tuple
```
```suggestion
            # Use two with statements inside each other as python 2.6 doesn't support passing a tuple
```

jfboismenu reviewed (approved):
Great work. Only two typos to fix. Good for QA! 😄

pscadding committed:
Apply suggestions from code review

Addressing CR comments, fixed the grammar in two comments.

Co-Authored-By: Jean-François Boismenu <jfboismenu@users.noreply.github.com>

pscadding committed:
updated the copyright year for the new files.

pscadding committed:
Fixed resource file.

pscadding committed:
Fixed the codecov badge.
Added a fallback str convert when six is not present.

pscadding committed:
Now doesn't use deprecated Qt method.

pscadding committed:
Bumped core version to `v0.19.1`

pscadding committed:
Fixed styling issue. Alpha was being passed to rgb color.

pscadding committed:
Styling; added the alpha back in, and set it to rgba instead.
